File: crawler/broker/send_to_broker.py
import logging
from typing import List

import requests
from conf.conf import conf
from models.models import Job, JobMeta, JobTbs

logger = logging.getLogger(__name__)


class Broker:
    def __init__(self, broker_uri):
        self.locator_uri = broker_uri

    def send_to_broker(self, data: List[Job], id) -> bool:
        try:
            final = []
            for d in data:
                url = ""
                state = True
                job_metadata = JobMeta(id=id, state=state, url=url)
                payload = JobTbs(metadata=job_metadata, data=d)
                final.append(payload.model_dump())

            broker_ingest_url = f"http://{conf.broker_url}:{conf.broker_http_port}/ingest"
            response = requests.post(
                broker_ingest_url, json=final, timeout=10
            )
            if response.status_code >= 400:
                logger.error(
                    f"broker returned status {response.status_code} for job {id}"
                )
                return False
            return True

        except requests.ConnectionError as e:
            logger.error(f"broker connection failed for job {id}: {e}")
            return False
        except requests.Timeout as e:
            logger.error(f"broker request timed out for job {id}: {e}")
            return False
        except Exception as e:
            logger.error(f"unexpected error sending to broker for job {id}: {e}")
            return False

File: server/internals/broker/send_to_broker.go
//here a queue will be made to send the jobs to the broker

package broker

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"time"

	"github.com/steel77-7/Web-Swab/config"
	"github.com/steel77-7/Web-Swab/internals/types"
)

func PushToBroker(job types.Job) error {
	client := &http.Client{
		Timeout: 5 * time.Second,
	}

	tbs, _ := json.Marshal([]types.JobTbs{
		types.JobTbs{
			Data: job,
			MetaData: types.Metadata{
				ID:    job.ID,
				Url:   config.Conf.SERVER_URL,
				State: false,
			},
		},
	},
	)
	log.Printf("pushing job %s to broker", job.ID)
	req, _ := http.NewRequest("POST", "http://"+config.Conf.BROKER_URL+":"+config.Conf.BROKER_PORT+"/ingest", bytes.NewBuffer(tbs))
	resp, err := client.Do(req)
	if err != nil {
		log.Printf("broker request failed: %v", err)
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode > 300 {
		log.Printf("broker returned status %d", resp.StatusCode)
		return fmt.Errorf("broker returned status %d", resp.StatusCode)
	}
	//just reply with something or some retry logic ............
	return nil
}
